### PR TITLE
[MI-2463] Added testcases for pipeline job request

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**

### CODE REVIEW

**Review by Kshitij-Katiyar — APPROVED** [review]

File: server/plugin/api_test.go [hunk 1]
```diff
@@ -960,3 +961,165 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
+				  "wrong: "mockOrganization",
+				  "projectName": "mockProjectName",
+				  "requestType": "mockRequestType"
+				}
+			  }`,
+			isPayloadInvalid: true,
+			statusCode:       http.StatusInternalServerError,
+		},
+		{
+			description: "HandlePipelineApproveOrRejectReleaseRequest: failed to approve/reject request due to some internal server error",
```
Comment (ayusht2810) [inline server/plugin/api_test.go:1078]:
```suggestion
			description: "HandlePipelineApproveOrRejectReleaseRequest: failed to approve/reject the request due to server error",
```

File: server/plugin/api_test.go [hunk 2]
```diff
@@ -960,3 +961,165 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
+				  "organization": "mockOrganization",
+				  "projectName": "mockProjectName",
+				  "requestType": "mockRequestType"
+				}
+			  }`,
+			updatePipelineApprovalRequestStatus: http.StatusBadRequest,
+			statusCode:                          http.StatusOK,
+		},
+		{
+			description: "HandlePipelineApproveOrRejectReleaseRequest: failed to approve/reject request and update the post",
```
Comment (ayusht2810) [inline server/plugin/api_test.go:1030]:
```suggestion
			description: "HandlePipelineApproveOrRejectReleaseRequest: failed to approve/reject the request and update the post",
```

File: server/plugin/api_test.go [hunk 3]
```diff
@@ -960,3 +961,165 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
+				  "projectName": "mockProjectName",
+				  "requestType": "mockRequestType"
+				}
+			  }`,
+			updatePipelineApprovalRequestStatus:    http.StatusBadRequest,
+			updatePipelineReleaseApprovalPostError: errors.New("failed to update post"),
+			statusCode:                             http.StatusInternalServerError,
+		},
+		{
+			description: "HandlePipelineApproveOrRejectReleaseRequest: failed to approve/reject request and fetch approval details",
```
Comment (ayusht2810) [inline server/plugin/api_test.go:1046]:
```suggestion
			description: "HandlePipelineApproveOrRejectReleaseRequest: failed to approve/reject the request and fetch approval details",
```

File: server/plugin/api_test.go (hunk 2 above)
Comment (ayusht2810) [reply server/plugin/api_test.go:1030]: fix everywhere

File: server/plugin/api_test.go [hunk 4]
```diff
@@ -960,3 +961,165 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
+				}
+			  }`,
+			updatePipelineApprovalRequestStatus:       http.StatusInternalServerError,
+			statusCode:                                http.StatusInternalServerError,
+			updatePipelineReleaseApprovalRequestError: errors.New("failed to update pipeline approval request"),
+		},
+	} {
+		t.Run(testCase.description, func(t *testing.T) {
+			mockAPI.On("LogError", testutils.GetMockArgumentsWithType("string", 3)...)
+			mockAPI.On("GetDirectChannel", mock.AnythingOfType("string"), mock.AnythingOfType("string")).Return(&model.Channel{}, nil)
```
Comment (ayusht2810) [inline server/plugin/api_test.go:1096]: use `GetMockArgumentsWithType`

File: server/plugin/client_test.go [hunk 5]
```diff
@@ -317,6 +317,80 @@ func TestCall(t *testing.T) {
...
+	defer monkey.UnpatchAll()
+	p := setupTestPlugin(&plugintest.API{})
+	for _, testCase := range []struct {
+		description string
+		err         error
+		statusCode  int
+	}{
+		{
+			description: "UpdatePipelineApprovalRequest: valid",
+			err:         nil,
```
Comment (ayusht2810) [inline server/plugin/client_test.go:330]:
```suggestion
```

File: server/plugin/client_test.go [hunk 6]
```diff
@@ -317,6 +317,80 @@ func TestCall(t *testing.T) {
...
+			statusCode:  http.StatusOK,
+		},
+		{
+			description: "UpdatePipelineApprovalRequest: with error",
+			err:         errors.New("mock-error"),
```
Comment (ayusht2810) [inline server/plugin/client_test.go:335]: change error

File: server/plugin/client_test.go [hunk 7]
```diff
@@ -317,6 +317,80 @@ func TestCall(t *testing.T) {
...
+	} {
+		t.Run(testCase.description, func(t *testing.T) {
+			monkey.PatchInstanceMethod(reflect.TypeOf(&client{}), "Call", func(_ *client, basePath, method, path, contentType, mattermostUserID string, inBody io.Reader, out interface{}, formValues url.Values) (responseData []byte, statusCode int, err error) {
+				return nil, testCase.statusCode, testCase.err
+			})
+
+			statusCode, err := p.Client.UpdatePipelineApprovalRequest(&serializers.PipelineApproveRequest{}, "mockOrganization", "mockProjectID", "mockMattermostUSerID", 1234)
+
+			if testCase.err != nil {
+				assert.Error(t, err)
```
Comment (ayusht2810) [inline server/plugin/client_test.go:347]: check for error value if possible

Comment (ayusht2810) [reply server/plugin/client_test.go:347]: check everywhere

File: server/plugin/client_test.go (hunk 6 above)
Comment (ayusht2810) [reply server/plugin/client_test.go:335]: check everywhere

File: server/plugin/client_test.go (hunk 5 above)
Comment (ayusht2810) [reply server/plugin/client_test.go:330]: check everywhere

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/api_test.go [hunk 8]
```diff
@@ -960,3 +961,165 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
+		},
+	} {
+		t.Run(testCase.description, func(t *testing.T) {
+			mockAPI.On("LogError", testutils.GetMockArgumentsWithType("string", 3)...)
+			mockAPI.On("GetDirectChannel", testutils.GetMockArgumentsWithType("string", 2)...).Return(&model.Channel{}, nil)
+			mockAPI.On("SendEphemeralPost", mock.AnythingOfType("string"), mock.AnythingOfType("*model.Post")).Return(&model.Post{})
+			mockAPI.On("UpdateEphemeralPost", mock.AnythingOfType("string"), mock.AnythingOfType("*model.Post")).Return(nil)
+
+			if !testCase.isPayloadInvalid {
+				mockedClient.EXPECT().UpdatePipelineApprovalRequest(gomock.Any(), "mockOrganization", "mockProjectName", "test-userID", 1234).Return(testCase.updatePipelineApprovalRequestStatus, testCase.updatePipelineReleaseApprovalRequestError)
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:1101]: the constants we have created in other branches use them here as well

File: server/plugin/api_test.go [hunk 9]
```diff
@@ -960,3 +961,165 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
+			monkey.PatchInstanceMethod(reflect.TypeOf(p), "UpdatePipelineReleaseApprovalPost", func(_ *Plugin, _, _, _ string) error {
+				return testCase.updatePipelineReleaseApprovalPostError
+			})
+
+			monkey.PatchInstanceMethod(reflect.TypeOf(p), "DM", func(_ *Plugin, _, _ string, _ bool, _ ...interface{}) (string, error) {
+				return "", nil
+			})
+
+			req := httptest.NewRequest(http.MethodGet, "/mockPath", bytes.NewBufferString(testCase.body))
+			req.Header.Add(constants.HeaderMattermostUserID, "test-userID")
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:1117]: same

File: server/plugin/api_test.go (hunk 8 above)
Comment (raghavaggarwal2308) [reply server/plugin/api_test.go:1101]: Created a ticket for this. I am doing the same thing across multiple PRs. Once all the PRs are merged, I will create a separate PR for such small enhancements.
https://brightscout.atlassian.net/browse/MI-2511?atlOrigin=eyJpIjoiNWFlOTMxODJjYWM4NDgwYWJiNjE2N2MxMDdjZDRkNjAiLCJwIjoiaiJ9

**Pull request merged**